### PR TITLE
Resolve chain projections.

### PR DESCRIPTION
**Pull request by ptersilie**
This PR implements resolving of chained projections. I also took the opportunity to clean up some code.

### CODE REVIEW

File: ykcompile/src/lib.rs [hunk 1]
```diff
@@ -1826,4 +1829,29 @@ mod tests {
...
+        inputs.3 = inputs.2.y;
+        let sir_trace = th.stop_tracing().unwrap();
+        let tir_trace = TirTrace::new(&*SIR, &*sir_trace).unwrap();
+        let ct = TraceCompiler::<&((usize, u8, usize), u8, S, usize)>::compile(tir_trace);
+
+        let t2 = (1usize, 2u8, 3usize);
+        let s2 = S { x: 5, y: 6 };
+        let mut args = (t2, 0u8, s2, 0usize);
+        ct.execute(&mut args);
+        assert_eq!(args.1, 2);
```
Comment (vext01) [inline ykcompile/src/lib.rs:1830]: We can make this test stronger by checking the whole inputs tuple is the right shape. That way we'd also check the read-only parts are intact.

File: ykcompile/src/lib.rs [hunk 2]
```diff
@@ -178,17 +178,17 @@ impl<TT> TraceCompiler<TT> {
     /// Takes a `Place`, resolves all projections, and returns a `Location` containing the result.
     fn resolve_projection(&mut self, p: &Place) -> Result<Location, CompileError> {
         let mut curloc = self.local_to_location(p.local)?;
-        if p.projection.len() > 1 {
-            todo!("Deal with remaining projections");
-        }
+        let mut base_ty = self.place_ty(&Place::from(p.local)).clone();
         for proj in &p.projection {
             // FIXME Get the type of a projection.
```
Comment (vext01) [inline ykcompile/src/lib.rs:184]: I think this comment is invalid now?

**Comment by vext01** [comment]
Looks good! Just minor stuff.

File: ykcompile/src/lib.rs (hunk 1 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:1830]: Good point.

File: ykcompile/src/lib.rs (hunk 2 above)
Comment (ptersilie) [reply ykcompile/src/lib.rs:184]: It is! Good spot.

**Comment by ptersilie** [comment]
Addressed comments.

**Comment by vext01** [comment]
Please squash.

**Comment by vext01** [comment]
And rebase.

**Comment by ptersilie** [comment]
Squashed and rebased.

**Pull request merged**